core/src/com/unciv/logic/battle/GreatGeneralImplementation.kt

```diff
@@ -0,0 +1,141 @@
+package com.unciv.logic.battle
+
+import com.unciv.logic.map.MapUnit
+import com.unciv.logic.map.TileInfo
+import com.unciv.models.ruleset.unique.StateForConditionals
+import com.unciv.models.ruleset.unique.Unique
+import com.unciv.models.ruleset.unique.UniqueType
+import com.unciv.logic.automation.SpecificUnitAutomation  // for Kdoc
+import com.unciv.logic.civilization.CivilizationInfo
+
+
+object GreatGeneralImplementation {
+    private data class GeneralBonusData(val general: MapUnit, val radius: Int, val filter: String, val bonus: Int) {
+        constructor(general: MapUnit, unique: Unique) : this(
+            general,
+            radius = unique.params[2].toIntOrNull() ?: 0,
+            filter = unique.params[1],
+            bonus = unique.params[0].toIntOrNull() ?: 0
+        )
+        @Deprecated("Remove with UniqueType.BonusForUnitsInRadius")
+        constructor(general: MapUnit) : this(general, 2, "All", 15)
+    }
+
+    /**
+     * Determine the "Great General" bonus for [unit] by searching for units carrying the [UniqueType.GreatGeneralAura] in the vicinity.
+     *
+     * Used by [BattleDamage.getGeneralModifiers].
+     *
+     * @return Percentage bonus as Int (typically 15), or 0 if no applicable Great General equivalents found 
+     */
+    fun getGreatGeneralBonus(unit: MapUnit): Int {
+        val civInfo = unit.civInfo
+        val nearbyGenerals = unit.getTile()
+            .getTilesInDistance(civInfo.maxGeneralBonusRadius)
+            .flatMap { it.getUnits() }
+            .filter { it.civInfo == civInfo && it.hasGreatGeneralUnique }
+        if (nearbyGenerals.none()) return 0
+
+        val greatGeneralModifier = nearbyGenerals
+            .flatMap { general ->
+                general.getMatchingUniques(UniqueType.GreatGeneralAura)
+                    .map { GeneralBonusData(general, it) } +
+                        general.getMatchingUniques(UniqueType.BonusForUnitsInRadius)
+                            .map { GeneralBonusData(general) }
+            }.filter {
+                // Support the border case when a mod unit has several
+                // GreatGeneralAura uniques (e.g. +50% as radius 1, +25% at radius 2, +5% at radius 3)
+                it.general.currentTile.aerialDistanceTo(unit.getTile()) <= it.radius
+                        && (it.filter == "All" || unit.matchesFilter(it.filter))
```

Comment:
Shouldn't `it.filter == All` also be tested in the `unit.matchesFilter(it.filter)`? Or is this a surprisingly large optimization?

Reply:
I think "All" is supported way down the call stack... BaseUnit.matchesFilter, and since the unmodded Great General should be efficient, yes, pure optimization. I remember leaving a comment - probably removed it thinking 'unnecessary'.